fix: use `lifespan` context manager in Starlette and FastAPI

cofin opened the pull request:
Modifes the Starlette and FastAPI integrations to use the `lifespan` context manager instead of the `startup`\`shutdown` hooks.  If the application already has a lifespan set, it is wrapped so that both execute.

cofin commented:
Closes #367

PokkaKiyo commented:
@cofin If you don't mind me adding, the changes you made to AdvancedAlchemy seem like it would run each config's on_startup/on_shutdown twice if the user does not define a custom lifespan, but only once if the user does define lifespan.

That's probably fine for create_all and engine.dispose, but maybe the add_event_handler functions in the config classes should be removed if this isn't intended?

cofin commented:
> @cofin If you don't mind me adding, the changes you made to AdvancedAlchemy seem like it would run each config's on_startup/on_shutdown twice if the user does not define a custom lifespan, but only once if the user does define lifespan.
> 
> That's probably fine for create_all and engine.dispose, but maybe the add_event_handler functions in the config classes should be removed if this isn't intended?
> 
> Also the outer shutdown doesn't run when hit with an exception 😅

Hah.  Is it obvious I don't use FastAPI daily? :)

cofin commented:
> @cofin If you don't mind me adding, the changes you made to AdvancedAlchemy seem like it would run each config's on_startup/on_shutdown twice if the user does not define a custom lifespan, but only once if the user does define lifespan.
> 
> That's probably fine for create_all and engine.dispose, but maybe the add_event_handler functions in the config classes should be removed if this isn't intended?
> 
> Also the outer shutdown doesn't run when hit with an exception 😅

Your suggestions were spot on.  I think I've *finally* addressed all of the issues on this one.  Thanks again.

PokkaKiyo commented:
Thank you!
